### ISSUE TITLE
[CT-2701] [Bug] Snapshot dbt_updated_at not updated

### ISSUE THREAD

**Issue by patkearns10** [dbt-labs/dbt-core#7869]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

This was brought up by a customer and I am kinda surprised I have never heard this before. 

[Our docs show](https://docs.getdbt.com/docs/build/snapshots)

two fields in the existing row should be updated when the `updated_at` timestamp is changed signifying a change of some value, the `dbt_valid_to` and the `dbt_updated_at`

![image](https://github.com/dbt-labs/dbt-core/assets/30663534/77a6c1db-3336-4abe-9a80-a8e398e16237)

Looks like only one update is happening and the `dbt_updated_at` does not get updated.

![Screen Shot 2023-06-14 at 2 06 07 PM](https://github.com/dbt-labs/dbt-core/assets/30663534/dd03040c-c348-45db-aae8-fec79da3a97e)

### Expected Behavior

According to the docs, the older row should have two fields updated:
`dbt_valid_to`
`dbt_updated_at`

### Steps To Reproduce

- create  snapshot with timestamp check
- add record `1 as id, '2023-01-01' as updated_at`
- run snapshot
- modify record `1 as id, '2023-01-02' as updated_at`
- run snapshot
- check the records in the output and see if `dbt_updated_at` has been modified in the original row

### Environment

```markdown
- OS:
- Python:
- dbt: tested 1.1 and 1.4
```

**Comment by dbeatty10** [comment]
Thanks for opening this @patkearns10!

See below for the reproducible example I'm using (with dbt-postgres) that shows the same thing that you are reporting.

But the thing I'm wondering: is this actually a problem that needs to be solved in the implementation of snapshots that use the `timestamp` strategy? Or would updating the documentation suffice?

<details>
<summary>

### Reprex</summary>

`snapshots/my_snapshot.sql`
```sql
{% snapshot my_snapshot %}

{{
    config(
      target_schema='snapshots',
      unique_key='id',
      strategy='timestamp',
      updated_at='updated_at',
    )
}}

{% if var("my_data_version", 1) == 1 %}

    select 1 as id, '2023-01-01'::timestamp as updated_at

{% else %}

    select 1 as id, '2023-01-02'::timestamp as updated_at

{% endif %}

{% endsnapshot %}
```

Take the 1st snapshot:
```shell
dbt snapshot --vars '{"my_data_version": 1}'
dbt show --inline "select * from {{ ref('my_snapshot') }}"
```

| id |          updated_at | dbt_scd_id           |      dbt_updated_at |      dbt_valid_from | dbt_valid_to |
| -- | ------------------- | -------------------- | ------------------- | ------------------- | ------------ |
|  1 | 2023-01-01 00:00:00 | 5aec37f35c393a7ef... | 2023-01-01 00:00:00 | 2023-01-01 00:00:00 |              |

Take a 2nd snapshot:

```
dbt snapshot --vars '{"my_data_version": 2}'
dbt show --inline "select * from {{ ref('my_snapshot') }}"
```

| id |          updated_at | dbt_scd_id           |      dbt_updated_at |      dbt_valid_from |        dbt_valid_to |
| -- | ------------------- | -------------------- | ------------------- | ------------------- | ------------------- |
|  1 | 2023-01-01 00:00:00 | 5aec37f35c393a7ef... | 2023-01-01 00:00:00 | 2023-01-01 00:00:00 | 2023-01-02 00:00:00 |
|  1 | 2023-01-02 00:00:00 | 8ff9567d22a84a2c9... | 2023-01-02 00:00:00 | 2023-01-02 00:00:00 |                     |

</details>

I think the key part of the implementation is below, which we can see _only_ updates the `dbt_valid_to` (but not `dbt_updated_at`) for rows that are no longer valid:
https://github.com/dbt-labs/dbt-core/blob/f767943fb22b493a762ff76c4d942eec7b781b2c/plugins/postgres/dbt/include/postgres/macros/materializations/snapshot_merge.sql#L5-L10

**Comment by patkearns10** [comment]
If we do nothing, we _should_ update the documentation.
If we want it to be, in my opinion, correct, then we should update the code and not update the documentation.

In my opinion it does not make sense that the `dbt_valid_to` would be greater timestamp than `dbt_updated_at`. The record was updated, so it should have a `dbt_updated_at` at that time. 

I do understand this is changing something that has been around for 4 years, so there is probably some concern there.

**Comment by dbeatty10** [comment]
Agreed that it's been implemented this way for 4+ years and it might do more harm than good to change it. At the very least, we'd need to understand all the implications extremely thoroughly before considering making such a change.

What row-level data point would a user want that they can't get right now?

Right now, it looks like `dbt_updated_at` basically means "created at" (when using the `check` strategy) and "earliest valid time" otherwise. If a user wants the latest time at which the row was "modified at", then this might do the trick:
```sql
coalesce(`dbt_valid_to`, `dbt_valid_from`)
```

If you are _always_ looking for a ["system time"](https://github.com/dbeatty10/Phippy-Goes-Fact-Finding#cross-walk-of-temporal-terminology) (a la SQL:2011), then that is not possible with the `timestamp` strategy, because that information is not always* captured and stored -- it is only captured and stored for the `check` strategy.

\*  When using [invalidate_hard_deletes](https://docs.getdbt.com/docs/build/snapshots#hard-deletes-opt-in) and a key is hard-deleted, `dbt_valid_to` will be a "system time" rather than an "application time".

I'm leaning heavily towards leaving the implementation as-is and updating the documentation.

We have a listing of gotchas related to snapshots listed [here](https://github.com/dbt-labs/dbt-core/discussions/7018), and the situation you are bringing up looks like it would another good entry for that list.

**Comment by patkearns10** [comment]
Right on, I understand the sentiment.

**Comment by ntsteph** [comment]
hi @patkearns10 and @dbeatty10,

thanks for this post, i had the same issue and fix this that way. 

however by overriding the macro snapshot_merge_sql in he dbt project, it looks like this affect all existing strategies (timestamp, check, custom strategy created).

Is it possible to change the default behaviour of the macro only for one strategy (for example the custom strategy created)  and not the others in he dbt project ?

Thanks in advance

**Comment by dbeatty10** [comment]
@ntsteph it sounds like you've created a custom snapshot strategy, and you're wondering if you can change the default behavior of `snapshot_merge_sql` for _just_ that custom snapshot strategy?

If so, you'd need to override the definition of the snapshot materialization for your adapter ([here](https://github.com/dbt-labs/dbt-core/blob/d91265411061463b42d6f4d6d94e094a42a96d2b/core/dbt/include/global_project/macros/materializations/snapshots/snapshot.sql#L65-L70) for dbt-postgres). So although it's _possible_, there's no "easy" way to override the logic for `snapshot_merge_sql` for a single snapshot strategy (but not the others).

**Comment by graciegoheen** [comment]
[From snapshot discussion:
](https://github.com/dbt-labs/dbt-core/discussions/7018)
6. **Problem 6**: When a row is no longer currently valid, `dbt_valid_to` is updated but `dbt_updated_at` is not (as described [here](https://github.com/dbt-labs/dbt-core/issues/7869)). This effectively means it is treated as a "row created at" rather than a "row updated at")
7. **Problem 7**: `dbt_updated_at` is a "system time" for the `check` strategy (a la [SQL:2011](https://github.com/dbeatty10/Phippy-Goes-Fact-Finding#cross-walk-of-temporal-terminology)), but it is an "application time" for the `timestamp` strategy.

- 💡 **Feature idea 9**: What if dbt either added a new column named `dbt_modified_at` or actually updated `dbt_updated_at` at the same time as `dbt_valid_to`?
- 💡 **Feature idea 10**: What if dbt treated `dbt_modified_at`/`dbt_updated_at` as a "system time" for both snapshot strategies (since `dbt_valid_from` & `dbt_valid_to` represent the "application time" for both strategies)?
- 💡 **Feature idea 11**: To complement `dbt_modified_at`, what if dbt added an immutable `dbt_created_at` column (representing the "system time" that the row was inserted)?
